node/router/src/lib.rs

```diff
@@ -84,6 +84,104 @@ const PEER_CACHE_FILENAME: &str = "cached_router_peers";
+    fn is_valid_peer_ip(&self, ip: SocketAddr) -> bool {
+        !self.is_local_ip(ip) && !is_bogon_ip(ip.ip()) && !is_unspecified_or_broadcast_ip(ip.ip())
+    }
+
+    /// Returns the maximum number of connected peers.
+    fn max_connected_peers(&self) -> usize {
+        self.tcp().config().max_connections as usize
+    }
+
+    /// Ensure we are allowed to connect to the given listener address of a peer.
+    ///
+    /// # Return Values
+    /// - `Ok(true)` if already connected (or connecting) to the peer.
+    /// - `Ok(false)` if not connected to the peer but allowed to.
+    /// - `Err(err)` if not allowed to connect to the peer.
+    fn check_connection_attempt(&self, listener_addr: SocketAddr) -> Result<bool> {
+        // Ensure the peer IP is not this node.
+        if self.is_local_ip(listener_addr) {
+            bail!("Dropping connection attempt to '{listener_addr}' (attempted to self-connect)");
+        }
+        // Ensure the node does not surpass the maximum number of peer connections.
+        if self.number_of_connected_peers() >= self.max_connected_peers() {
+            bail!("Dropping connection attempt to '{listener_addr}' (maximum peers reached)");
+        }
+        // Ensure the node is not already connected to this peer.
+        if self.is_connected(listener_addr) {
+            debug!("Dropping connection attempt to '{listener_addr}' (already connected)");
+            return Ok(true);
+        }
+        // Ensure the node is not already connecting to this peer.
+        if self.is_connecting(listener_addr) {
+            debug!("Dropping connection attempt to '{listener_addr}' (already connecting)");
+            return Ok(true);
+        }
+        Ok(false)
+    }
+
+    /// Attempts to connect to the given peer's listener address.
+    ///
+    /// Returns None if we are already connected to the peer or cannot connect.
+    /// Otherwise, it returns a handle to the tokio tasks that sets up the connection.
+    fn connect(&self, listener_addr: SocketAddr) -> Option<JoinHandle<bool>> {
+        // Return early if the attempt is against the protocol rules.
+        match self.check_connection_attempt(listener_addr) {
+            Ok(true) => return None,
+            Ok(false) => {}
+            Err(error) => {
+                warn!("{error}");
+                return None;
+            }
+        }
+
+        // Determine whether the peer is trusted or a bootstrap node in order to decide
+        // how problematic any potential connection issues are.
+        let is_trusted_or_bootstrap =
+            self.is_trusted(listener_addr) || bootstrap_peers::<N>(false).contains(&listener_addr);
+
+        let tcp = self.tcp().clone();
+        Some(tokio::spawn(async move {
+            debug!("Connecting to {listener_addr}...");
+            // Attempt to connect to the peer.
+            match tcp.connect(listener_addr).await {
```

Comment:
Do we keep candidate peers in this case?

Reply:
if a connection is successful, then even if a peer had been unknown until then, it is retained as a future candidate after a disconnect